CONTRIBUTING.md

```diff
@@ -0,0 +1,211 @@
+tox run -e integration   # integration tests
+tox                      # runs 'lint', 'unit', 'static', and 'coverage-report' environments
+```
+
+* More on [Formatting]()
+* More on [Linting]()
+* More on [Unit Testing]()
+* More on [Integration Testing]()
+
+## Building the charms
+
+In this repository, you'll find two machine charms.
+The k8s charm handles the deployment of the control plane node, and the k8s-worker charm takes care of deploying worker nodes.
+Build all the charms in this git repository using:
+
+```shell
+charmcraft pack -p charms/worker
+charmcraft pack -p charms/worker/k8s
+```
+
+## Tox Environments
+
+### Formatting
+
+This repo uses `isort` and `black` to format according to rules setup in `./charms/worker/k8s/pyproject.yaml`.
+
+Running the formatter is as easy as:
+
+```shell
+tox run -e format
+```
+
+If the github CI is complaining about invalid formatting, it could be due to an updated version of black. To fix locally, just run tox with a refreshed environment. The pip requirements will be refreshed locally on your machine and the formatter should be adjusted.
+
+```shell
+tox run -re format
+```
+
+### Linting
+
+This repo uses static analysis tools configured with `./charms/worker/k8s/pyproject.yaml` to ensure that all source files maintain a similar code style and docs style.
+
+Running the linter is as easy as:
+
+```shell
+tox run -e lint,static
+```
+
+If the github CI is complaining about invalid linting, it could be due to an updated version of one of the linter tools. To fix locally, just run tox with a refreshed environment. The pip requirements will be refreshed locally on your machine and the linters should be adjusted.
+
+```shell
+tox run -re lint,static
+```
+
+### Unit Testing
+
+This repo uses `pytest` to execute unit tests against the charm code, and create a coverage report after the unit tests are completed. The unit tests are defined in `./charms/worker/k8s/tests/unit/`
+
+Running the unit tests are as easy as:
+
+```shell
+tox run -e unit,coverage-report
+```
+
+Since the same charm code is executed on the worker and control-plane, in some unit test modules, we'll parameterize the tests to run against both the worker and control-plane to confirm both paths are tested. See `./charms/worker/k8s/tests/unit/test_base.py` for examples.
+
+### Integration Testing
+
+This repo uses `pytest` and `pytest-operator` to execute functional/integration tests against the charm files. The integration tests are defined in `./tests/integration`. Because this repo consists of two charms, the integration tests will build two charm files automatically without you doing anything. If you want to use specific charm files, just make sure the `.charm` files are in the top-level paths and the integration tests will find them if they are named appropriately (eg `./k8s-worker_*.charm` or `k8s_*.charm`). The charms are deployed according to the bundle defined in `./tests/integration/test-bundle.yaml`.
```

Comment:
```suggestion
This repo uses `pytest` and `pytest-operator` to execute functional/integration tests against the charm files. The integration tests are defined in `./tests/integration`. Because this repo consists of two charms, the integration tests will build two charm files automatically without you doing anything. If you want to use specific charm files, just make sure the `.charm` files are in the top-level paths and the integration tests will find them if they are named appropriately (eg `./*.charm`). The charms are deployed according to the bundle defined in `./tests/integration/test-bundle.yaml`.
```

I don't think we've got any integration tests yet in this repo...

Reply:
nah, it's fine